Handling multiple concurrent probe users.

derek0883 opened the pull request:
Handle multiple concurrent probe users, fixed issue #872.
Event naming pattern changed to $(eventname)_bcc_$(pid)
Detect /sys/kernel/debug/tracing/instances in bpf_attach_probe,
if it exist, then will create per-instance event, if failed then make global
event instance as same as before.

4ast commented:
lgtm
[buildbot, ok to test]
@drzaeus77 @palmtenor @markdrayton @goldshtn @brendangregg 
please review.
I think it's a good change and I don't expect a fallout from it,
but I may not have thought of all cases.

derek0883 commented:
@4ast @brendangregg 
I have fixed the auto testing failure. :)

goldshtn commented:
I'm just wondering: why can't the logic for creating the "_bcc_$PID" pattern be on the C side, so that we don't have to change the external API at all? Right now it looks like we're going to have to add this for C++, Python, and Lua (BTW I don't see Lua on this PR).

derek0883 commented:
@goldshtn 
That because when C code bpf_attach_kprobe get called, the argument is:
event_desc = p:kprobes/p_sys_open sys_open
event = sys_open
Front-end library will generate event_desc based on event name.
technically I can do that in C, But If I do that in C code,  I need replace
```
p:kprobes/p_sys_open           sys_open  
p:kprobes/p_sys_open_bcc_$PID  sys_open
```
I feel that This change will make that piece of code a little weird. 
I also thought that I can ignore event_desc, just do something like this:
snprintf(new_desc, "p:kprobe/p_%s_bcc_%d %s", event, getpid(), event);
Then I thought ignore a argument without front-end notice is not good neither. 
And we add _bcc_$PID in front-end library, not end-user script writer, this is 1 time change. 
Do you think this is acceptable?
Yes, I forgot to change LUA script, I will do that later tonight. thanks.

goldshtn commented:
Well, if we're making a change to the BCC library API anyway, maybe we can get rid of the `event_desc` and have the C side generate that? @4ast @drzaeus77

drzaeus77 commented:
> Well, if we're making a change to the BCC library API anyway, maybe we can get rid of the event_desc and have the C side generate that? @4ast @drzaeus77

That sounds reasonable to me. Projects that link against bcc (e.g. gobpf) will have to adjust but I think that's ok.

derek0883 commented:
> Well, if we're making a change to the BCC library API anyway, maybe we can get rid of the event_desc and have the C side generate that? 

So in bpf_attach_probe, remove event_desc argument, right?  This is easier for kprobe.
I just tried to make this change, I almost finished bpf_attach_kprobe, 
while I worked on bpf_attach_uprobe, I realized that bpf_attach_uprobe might want keep event_desc. 
e.g. in src/cc/BPF.cc attach_uprobe.  

```c++
  std::string probe_event_desc = attach_type_prefix(attach_type);
  probe_event_desc += ":uprobes/" + probe_event + " ";
  probe_event_desc += binary_path + ":0x" + uint_to_hex(sym.offset);
  void* res =
      bpf_attach_uprobe(probe_fd, probe_event.c_str(), probe_event_desc.c_str(),
                        pid, cpu, group_fd, cb, cb_cookie);
```
without probe_event_desc,  We need pass binary_path and sym.offset to bpf_attach_uprobe in libbpf.c

currently in src/cc/libbpf.c, kprobe and uprobe interface are same.  If we make this change, 
```c
void * bpf_attach_kprobe(int progfd, const char *event,
                        const char *event_desc,
                        pid_t pid, int cpu, int group_fd,
                        perf_reader_cb cb, void *cb_cookie) {
  return bpf_attach_probe(progfd, event, event_desc, "kprobe", pid, cpu, group_fd, cb, cb_cookie);
}

void * bpf_attach_uprobe(int progfd, const char *event,
                        const char *event_desc, 
                        pid_t pid, int cpu, int group_fd,
                        perf_reader_cb cb, void *cb_cookie) {
  return bpf_attach_probe(progfd, event, event_desc, "uprobe", pid, cpu, group_fd, cb, cb_cookie);
}   

1. either remove event_desc for kprobe, keep event_desc for uprobe
2. or remove event_desc for both, and pass binary_path and sym.offset for uprobe only.

this will make the interface different.  Maybe keep both interface same is better for front-end user ?
Another reason to keep in front-end is that, maybe kernel will add some more syntax for it,something like dynamic_debug the kernel current have. e.g.

```
# echo 'file svcsock.c line 1603 +p'  >  <debugfs>/dynamic_debug/control
```
line could be a line number or rang, e.g. 10-100. 
keep it in front-end, so user can change python script debug it without rebuild C lib?  :)

derek0883 commented:
@vmg 
Hi Vincent 
Would you please help me take a look how to make this changes for LUA.
I think the discussing and decision about  event_desc takes time. I'd like to make lua part for current changes. the change is simple. 
1. Add _bcc_$PID for ev_name.
2. for detach kprobe, currently pass event_desc, we need pass two argument,  event_desc, and ev_name.
I tried the changes by myself. I only added 1 line to  src/lua/bcc/bpf.lua
```
local posix = require("posix")
``` 
After rebuild 
```bash
build$ sudo ./src/lua/bcc-lua ../tools/biosnoop.lua
./src/lua/bcc-lua: attempt to write to undeclared variable posix
stack traceback:
	[C]: in function 'error'
	bcc.lua:1838: in function <bcc.lua:1837>
	[C]: at 0x7fd16c351ad0
	[C]: in function 'require'
	/usr/share/lua/5.1/posix.lua:2: in main chunk
	[C]: in function 'require'
	bcc.lua:1968: in function <bcc.lua:1946>
	[C]: in function 'require'
	bcc.lua:5776: in function <bcc.lua:5736>
	[C]: at 0x00793580
```
Would you please help me make the changes. :) Thanks!

4ast commented:
> need pass binary_path and sym.offset to bpf_attach_uprobe in libbpf.c?

yes. i think that's a good plan. Things like ":0x" + uint_to_hex(sym.offset) should
ideally be hidden within libbpf.c

derek0883 commented:
@4ast 
Thanks for confirming that!  Then I'm going to remove event_desc for uprobe too, generate it inside libbpf.c. :)

brendangregg commented:
I'm back online. FWIW, I tested the current code and it works. :)

4ast commented:
@derek0883 you're still working on moving suffix gen into libbpf.c, I assume, right?

derek0883 commented:
@brendangregg Thanks for confirm that.
@4ast  Yes, I'm going to send another PR for review this weekend. 
This version made more changes than previous one. :)

derek0883 commented:
@4ast I submitted new fix for review.  
Moved event desc to libbpf.c, So front-end don't need to handle that.  But changed C API.  :)

4ast commented on src/cc/BPF.h:
```diff
@@ -29,11 +29,6 @@
 
 namespace ebpf {
 
-enum class bpf_attach_type {
```
could you keep it as 'enum bpf_attach_type' inside libbpf.h instead of introducing #define -s ?
compiler unfortunately consider int == enum for argument checking purpose, but it's still cleaner api.

4ast commented on src/cc/BPF.h:
```diff
@@ -111,31 +106,31 @@ class BPF {
-    switch (type) {
-    case bpf_attach_type::probe_entry:
+  std::string attach_type_debug(int attach_type) {
+    switch (attach_type) {
```
for example, by keeping enum compiler will be able to warn here that some enum values are not handled in the switch...

4ast commented on src/cc/libbpf.c:
```diff
@@ -336,14 +338,18 @@ static int bpf_attach_tracing_event(int progfd, const char *event_path,
+{
   int kfd;
-  char buf[256];
+  char buf[strlen(ev_name)+256];
```
i'm not sure what this 'strlen(ev_name)+256' math suppose to do.
if i understand it correctly, the size of this array is still a rough guess of how much buffer space is needed.
if so, just statically set it to 256. all ev_names suppose to be short?

4ast commented on src/cc/libbpf.c:
```diff
@@ -355,52 +361,98 @@ static void * bpf_attach_probe(int progfd, const char *event,
+    snprintf(buf, sizeof(buf), "/sys/kernel/debug/tracing/instances/%s", new_name);
+    if (mkdir(buf, 0755) == -1) 
+      goto retry;
+    snprintf(buf, sizeof(buf), "/sys/kernel/debug/tracing/instances/%s/events/%ss/%s", 
```
this one is probably the largest printed string, so makes to check the value of snprintf for overflow.

4ast commented on src/cc/libbpf.c:
```diff
@@ -355,52 +361,98 @@ static void * bpf_attach_probe(int progfd, const char *event,
+                        perf_reader_cb cb, void *cb_cookie) 
+{
+  int kfd;
+  char buf[strlen(binary_path)+strlen(ev_name)+256];
```
same here. imo 256 + snprintf overflow check would be cleaner. no need to check all snprintfs only the largest.

vmg commented:
Hey @derek0883! Sorry I missed your ping. I don't see the `local posix = require("posix")` line in the diff. Could you please push the broken code so I can take a look at it? Thank you!

derek0883 commented:
@4ast I have submitted new changes for review. 
1.  keep enum bpf_attach_type in libbpf.h, but I renamed it to bpf_probe_attach_type, as we have another "enum bpf_attach_type" in  src/cc/compat/linux/bpf.h. this will cause compiling error for b_frontend_action.cc.
2. using static buf size plus check for snprintf. In bpf_attach_uprobe, as binary_path is not fixed, 256 might not enough, especially for non-english system. So I set the size to linux pre-defined macro PATH_MAX.
Feel free to let me know if you have other suggestion :).  Thanks.

@vmg 
Hi vincent, Thanks for your response. I have moved _bcc_pid stuff to libbpf.c, So front-end side such as python or LUA don't need to handle that. So I don't have that require("posix") question any more. :)

vmg commented:
:+1: Fantastic.

4ast commented on src/cc/BPF.h:
```diff
@@ -56,23 +51,23 @@ class BPF {
   StatusTuple attach_kprobe(
       const std::string& kernel_func, const std::string& probe_func,
-      bpf_attach_type attach_type = bpf_attach_type::probe_entry,
+      bpf_probe_attach_type = probe_entry,
```
how does this compile? or this is a github diff view bug?

4ast commented on src/cc/libbpf.h:
```diff
@@ -24,6 +24,11 @@
 #endif
 
+enum bpf_probe_attach_type {
+	probe_entry,
```
please make it upper case. Like: BPF_PROBE_ENTRY
to reduce the chance of conflicts.

4ast commented on tools/argdist.py:
```diff
@@ -20,7 +20,7 @@
         next_probe_index = 0
         streq_index = 0
-        aliases = {"$PID": "bpf_get_current_pid_tgid()"}
+        aliases = {"$PID": "(bpf_get_current_pid_tgid() >> 32)"}
```
hmm. a merge conflict? this patch already applied.

derek0883 commented:
@4ast 
1. Are you refer to :: in the bpf_attach_type::probe_entry?  As I moved the enum to libbpf.h, It protected by extern "C" {,  I removed the class key word. So we don't need :: to access the member. 
2. Changed it to BPF_PROBE_ENTRY and BPF_PROBE_RETURN.
3. That is a conflict, I I saw the page shows "This branch is out-of-date with the base branch". So I did  a "git rebase".  turns out not necessary.  Now no conflict.  
Feel free to let me know if you have other suggestion :). Thanks.

brendangregg commented:
many thanks!
